storage/ndb/rest-server/data-access-rondb/src/rdrs_rondb_connection.cpp

```diff
@@ -0,0 +1,405 @@
+  memcpy(__instance->node_ids, node_ids, node_ids_len * sizeof(Uint32));
+  __instance->node_ids_len = node_ids_len;
+
+  __instance->connection_retries            = connection_retries;
+  __instance->connection_retry_delay_in_sec = connection_retry_delay_in_sec;
+
+  __instance->ndbConnection      = nullptr;
+  __instance->reconnectionThread = nullptr;
+
+  int retCode = 0;
+  retCode     = ndb_init();
+  if (retCode != 0) {
+    return RS_SERVER_ERROR(ERROR_001 + std::string(" RetCode: ") + std::to_string(retCode));
+  }
+
+  return __instance->Connect();
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RS_Status RDRSRonDBConnection::Connect() {
+
+  LOG_INFO(std::string("Connecting to ") + connection_string);
+
+  {
+    std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+    if (stats.is_shutdown) {
+      return RS_SERVER_ERROR(ERROR_037);
+    }
+    require(stats.connection_state != CONNECTED);
+  }
+
+  {
+    std::lock_guard<std::mutex> guard(connectionMutex);
+    require(ndbConnection == nullptr);
+
+    int retCode   = 0;
+    ndbConnection = new Ndb_cluster_connection(connection_string, node_ids[0]);
+    retCode       = ndbConnection->connect(connection_retries, connection_retry_delay_in_sec, 0);
+    if (retCode != 0) {
+      return RS_SERVER_ERROR(ERROR_002 + std::string(" RetCode: ") + std::to_string(retCode));
+    }
+
+    retCode = ndbConnection->wait_until_ready(30, 0);
+    if (retCode != 0) {
+      return RS_SERVER_ERROR(ERROR_003 + std::string(" RetCode: ") + std::to_string(retCode) +
+                             std::string(" Lastest Error: ") +
+                             std::to_string(__instance->ndbConnection->get_latest_error()) +
+                             std::string(" Lastest Error Msg: ") +
+                             std::string(__instance->ndbConnection->get_latest_error_msg()));
+    }
+  }
+
+  {
+    std::lock_guard<std::mutex> guardInfo(connectionInfoMutex);
+    stats.connection_state = CONNECTED;
+  }
+
+  LOG_INFO("RonDB connection and object pool initialized");
+  return RS_OK;
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RS_Status RDRSRonDBConnection::GetInstance(RDRSRonDBConnection **rdrsRonDBConnection) {
+  if (__instance == nullptr) {
+    LOG_ERROR(ERROR_035);
+    return RS_SERVER_ERROR(ERROR_035);
+  }
+
+  std::lock_guard<std::mutex> guardInfo(__instance->connectionInfoMutex);
+  if (__instance->ndbConnection == nullptr || __instance->stats.connection_state != CONNECTED ||
+      __instance->stats.is_shutdown) {
+    LOG_ERROR(ERROR_033);
+    return RS_SERVER_ERROR(ERROR_033);
+  }
+
+  *rdrsRonDBConnection = __instance;
+  return RS_OK;
+}
+
+//--------------------------------------------------------------------------------------------------
+
+RS_Status RDRSRonDBConnection::GetNdbObject(Ndb **ndb_object) {
```

Comment:
Looks like a potential Concurrency bottleneck. Probably could do with some split of mutexes.